Add access docs for Ownable (#352)
OpenZeppelin/cairo-contracts: docs/Access.md

# Access

> Expect these modules to evolve.

Access control—that is, "who is allowed to do this thing"—is incredibly important in the world of smart contracts. The access control of your contract may govern who can mint tokens, vote on proposals, freeze transfers, and many other things. It is therefore critical to understand how you implement it, lest someone else [steals your whole system](https://blog.openzeppelin.com/on-the-parity-wallet-multisig-hack-405a8c12e8f7/).

## Table of Contents

* [Ownable](#ownable)
  * [Quickstart](#quickstart)
  * [Ownable library functions](#ownable-library-functions)
    * [`assert_only_owner`](#assert_only_owner)
    * [`owner`](#owner)
    * [`transfer_ownership`](#transfer_ownership)
    * [`renounce_ownership`](#renounce_ownership)
  * [Ownable events](#ownable-events)
    * [`OwnershipTransferred`](#ownershiptransferred)

## Ownable

The most common and basic form of access control is the concept of ownership: there’s an account that is the `owner` of a contract and can do administrative tasks on it. This approach is perfectly reasonable for contracts that have a single administrative user.

> OpenZeppelin Contracts provides Ownable for implementing ownership in your contracts.

Reviewer: I'd add a link to the contract or docs, no?

### Quickstart

> Setup simply consists of assigning an owner within the constructor of a Cairo contract like this:

Reviewer: ```suggestion
This library's initializer simply consists of assigning an owner within the constructor of a Cairo contract like this:
```
I'd push extensibility nomenclature and concepts throughout the docs, we cannot expect people to have read other sections so we should at least aid their learning.

```cairo
from openzeppelin.access.ownable import Ownable

@constructor
func constructor{
        syscall_ptr : felt*,
        pedersen_ptr : HashBuiltin*,
        range_check_ptr
    }(owner: felt):
    Ownable.initializer(owner)
    return ()
end
```

> To prevent non-owners from calling a function, add in the `assert_only_owner` method like this:

Reviewer: ```suggestion
To restrict a function's access to the owner only, add in the `assert_only_owner` method like this:
```
i think this reads better

```cairo
from openzeppelin.access.ownable import Ownable

func protected_function{
        syscall_ptr : felt*,
        pedersen_ptr : HashBuiltin*,
        range_check_ptr
    }():
    Ownable.assert_only_owner()
    return ()
end
```

### Ownable library functions

```cairo
func assert_only_owner():
end

func owner() -> (owner: felt):
end

func transfer_ownership(new_owner: felt):
end

func renounce_ownership():
end
```

#### `assert_only_owner`

> Throws if called by any account other than the owner.

Reviewer: ```suggestion
Reverts if called by any account other than the owner.
```
i think this is the right term

Parameters:

None.

Returns:

None.

#### `owner`

Returns the address of the current owner.

Parameters:

None.

Returns:

```cairo
owner: felt
```

#### `transfer_ownership`

Transfers ownership of the contract to a new account (`new_owner`). Can only be called by the current owner.

Emits a [`OwnershipTransferred`](#ownershiptransferred) event.

Parameters:

```cairo
new_owner: felt
```

Returns:

None.

#### `renounce_ownership`

> Leaves the contract without owner. It will not be possible to call onlyOwner functions anymore. Can only be called by the current owner.

Reviewer: ```suggestion
Leaves the contract without owner. It will not be possible to call functions with `assert_only_owner` anymore. Can only be called by the current owner.
```

end

func transfer_ownership(new_owner: felt):
end

func renounce_ownership():
end

func _transfer_ownership(new_owner: felt):
end
```

#### `initializer`

Initializes Ownable access control and should be called in the implementing contract's constructor. Assigns `owner` as the initial owner address of the contract.

This must be called only once.

Parameters:

```cairo
owner: felt
```

Returns:

None.

#### `assert_only_owner`

Reverts if called by any account other than the owner.

Parameters:

None.

Returns:

None.

#### `owner`

Returns the address of the current owner.

Parameters:

None.

Returns:

```cairo
owner: felt
```

#### `transfer_ownership`

Transfers ownership of the contract to a new account (`new_owner`). Can only be called by the current owner.

Emits a [`OwnershipTransferred`](#ownershiptransferred) event.

Parameters:

```cairo
new_owner: felt
```

Returns:

None.

#### `renounce_ownership`

Leaves the contract without owner. It will not be possible to call functions with `assert_only_owner` anymore. Can only be called by the current owner.

Emits a [`OwnershipTransferred`](#ownershiptransferred) event.

Parameters:

None.

Returns:

None.

> 
> <h4 id="transfer-ownership-internal"><code>_transfer_ownership</code></h4>
> 
> Transfers ownership of the contract to a new account (`new_owner`). [Internal function](./Extensibility.md#libraries) without access restriction.

Author: ```suggestion

<h4 id="transfer-ownership-internal"><code>_transfer_ownership</code></h4>

Transfers ownership of the contract to a new account (`new_owner`). Unprotected method without access restriction.
```

@martriay Thoughts on calling this as an `unprotected` method since it's within the namespace (following the [Extensibility Pattern](https://github.com/OpenZeppelin/cairo-contracts/blob/main/docs/Extensibility.md#libraries), `internal` funcs should not be within the namespace) or should we move it outside?

Reviewer: I guess the underscore today means unprotected, and internal functions should be outside the namespace. For consistency with the Solidity lib, I'd leave it as part of the API, no strong opinions on naming. Interested in @Amxx's thoughts.

func renounce_ownership():
end

func _transfer_ownership(new_owner: felt):
end
```

#### `initializer`

Initializes Ownable access control and should be called in the implementing contract's constructor.

This can only be called once.

Parameters:

```cairo

> owner:felt

Author: That's fair. I think adding a line to the `initializer`'s description should help clarify. All of the other method descriptions seem pretty explicit regarding what `owner` is

None.

<h4 id="transfer-ownership-internal"><code>_transfer_ownership</code></h4>

> Transfers ownership of the contract to a new account (`new_owner`). Internal function without access restriction.

Reviewer: I'd link to the extensibility docs where `internal` gets defined
